### PR TITLE
Add asset manifest parsing benchmark

### PR DESCRIPTION
**Pull request by andrewkolos**
This change adds a benchmark for loading and parsing the asset manifest from storage. It does this be mimicking the [implementation of AssetImage.obtainKey](https://github.com/flutter/flutter/blob/7ded3d91daf9478126b902b9508f1aeb9fb91e3a/packages/flutter/lib/src/painting/image_resolution.dart#L287), using AssetBundle.loadStructuredData to load and parse the file. It clears the bundle's cache between iterations.

Running this locally on an emulated Pixel 5 (API 33), I noticed some interesting results:

* 1 iteration: ~50ms
* 10 iterations: ~200ms
* 100 iterations: ~950 ms
* 1000 iterations: ~2900 ms

The runtime does not vary linearly with iteration count. Perhaps this is due to JIT optimization. Otherwise, I missed something.

~~TODO: I'm considering exposing the `_manifestParser` function as it's now duplicated in at least two tests.~~

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

File: dev/benchmarks/microbenchmarks/lib/foundation/load_structured_data.dart [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+
+  final BenchmarkResultPrinter printer = BenchmarkResultPrinter();
+  WidgetsFlutterBinding.ensureInitialized();
+  final Stopwatch watch = Stopwatch();
+  final PlatformAssetBundle bundle = PlatformAssetBundle();
+
+  watch.start();
+  for (int i = 0; i < _kNumIterations; i++) {
+    bundle.clear();
+    await bundle.loadStructuredData('gpay_asset_manifest.json', _manifestParser);
```
Comment (jonahwilliams) [inline dev/benchmarks/microbenchmarks/lib/foundation/load_structured_data.dart:26]: nit: call this "money_asset_manifest"

Comment (jonahwilliams) [reply dev/benchmarks/microbenchmarks/lib/foundation/load_structured_data.dart:26]: (here and elsewhere :) )

**Comment by jonahwilliams** [comment]
release mode is AOT, so its likely not due to JIT. Instead it could be a mix of memory locality, scheduling (due to the async nature of asset loading), and garbage collection. Its reasonable that the cost varies per count.

If you wanted to rule out cost of loading the asset from memory, you could instead cache the raw bytes and measure the cost of utf8 conversion followed by JSON parsing.

**Comment by andrewkolos** [comment]
> If you wanted to rule out cost of loading the asset from memory, you could instead cache the raw bytes and measure the cost of utf8 conversion followed by JSON parsing.

I just made this change, and the more I think about it, the more it makes sense in that it most closely tests the decoding+parsing change we are looking to make. Still, one thing this neglects is that the binary encoding may result in faster load from storage since the file will be smaller. Do you think this might be worth putting back into the test loop? The thing I wouldn't like about this is, if we ever changed the iteration count in the future, it would invalidate* previous data due to storage locality--though I do wonder if changing the iteration count invalidates data regardless.

*By "invalidate", I mean that it would no longer be useful to compare metrics from before the change to after the change.

**Comment by jonahwilliams** [comment]
I think it depends on how much variation you see when running on a real device. If the numbers are fairly stable around ~30 iterations, I think its fine if it doesn't scale completely linearly. We only need to worry about the delta between these numbers and the updated numbers

**Review by jonahwilliams — APPROVED** [review]
LGTM

**Pull request merged**